streaming/src/main/scala/org/apache/spark/streaming/receiver/BlockGenerator.scala

```diff
@@ -227,16 +227,21 @@ private[streaming] class BlockGenerator(
   def isStopped(): Boolean = state == StoppedAll
 
   /** Change the buffer to which single records are added to. */
-  private def updateCurrentBuffer(time: Long): Unit = synchronized {
+  private def updateCurrentBuffer(time: Long): Unit = {
     try {
-      val newBlockBuffer = currentBuffer
-      currentBuffer = new ArrayBuffer[Any]
-      if (newBlockBuffer.size > 0) {
-        val blockId = StreamBlockId(receiverId, time - blockIntervalMs)
-        val newBlock = new Block(blockId, newBlockBuffer)
-        listener.onGenerateBlock(blockId)
+      var newBlock: Block = null
+      synchronized {
+        if (currentBuffer.nonEmpty) {
+          val newBlockBuffer = currentBuffer
+          currentBuffer = new ArrayBuffer[Any]
+          val blockId = StreamBlockId(receiverId, time - blockIntervalMs)
+          listener.onGenerateBlock(blockId)
+          newBlock = new Block(blockId, newBlockBuffer)
+        }
+      }
+
+      if (newBlock != null) {
```

Comment:
The condition regarding state change can be done. But did you suggest that to solve the race condition? Because I dont think it solves it. 

I dont think that the timer thread that calls `updateCurrentBuffer` will be interrupted as the timer is stopped with [interruptTimer = false](https://github.com/apache/spark/blob/master/streaming/src/main/scala/org/apache/spark/streaming/receiver/BlockGenerator.scala#L145).